File: api-server/src/main/java/com/mgu/reactive/tutorial/PersonController.java
package com.mgu.reactive.tutorial;

import java.time.Duration;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import org.springframework.web.bind.annotation.*;

import reactor.core.publisher.Mono;

@RestController
public class PersonController {

    private Map<Long, Person> people = new HashMap<>();
    private Map<Long, String> domains = new HashMap<>();

    public PersonController() {
        people.put(1L, new Person(1L, "userX"));
        people.put(2L, new Person(2L, "userY"));
        people.put(3L, new Person(3L, "userZ"));
        domains.put(1L, "Darts");
        domains.put(2L, "Spring");
        domains.put(3L, "Kube");
    }
    
    @GetMapping("/persons")
    public Set<Long> getPersonIds() {
        return people.keySet();
    }

    @GetMapping("/persons/{id}")
    public Person getPerson(@PathVariable Long id, @RequestParam(name = "delay", required = false) String delay) {
        if (delay!=null) {
            try {
                Thread.sleep(Integer.parseInt(delay)*1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        return people.get(Long.valueOf(id));
    }

    @GetMapping("/persons/{id}/domain")
    public Mono<String> getPersonDomains(@PathVariable Long id, @RequestParam(name = "delay",required = false) String delay) {
        return Optional.ofNullable (delay)
            .map ( d -> Mono.delay (Duration.ofSeconds(Integer.parseInt(d)*2))
                .thenReturn (domains.get (Long.valueOf(id))))
            .orElse (Mono.just (domains.get(Long.valueOf(id))));
    }
}

record Person(Long id, String name) { }

File: api-server-kotlin/src/main/kotlin/com/mgu/reactive/tutorial/PersonController.kt
package com.mgu.reactive.tutorial

import kotlinx.coroutines.delay
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
class PersonController {
    val people = mapOf(
            Integer.valueOf(1) to Person(1,"Marc"),
            Integer.valueOf(2) to Person(2,"Eric"),
            Integer.valueOf(3) to Person(3,"Amélie")
    )

    val domains = mapOf(
            Integer.valueOf(1) to "Darts",
            Integer.valueOf(2) to "Spring",
            Integer.valueOf(3) to "Kube"
    )

    @GetMapping("/person")
    fun getPersonIds() = people.keys

    @GetMapping("/person/{id}")
    suspend fun getPerson(@PathVariable id :Integer, @RequestParam(name="delay", required = false) delay: String): Person? {
        if(delay!=null) {
            delay(delay.toLong() * 1000)
        }
        return people.get(id.toInt())
    }

    @GetMapping("/person/{id}/domain")
    suspend fun getPersonDomain(@PathVariable id :Integer, @RequestParam(name="delay", required = false) delay: String): String? {
        if(delay!=null) {
            delay(delay.toLong() * 2000)
        }
        return domains.get(id.toInt())
    }
}

data class Person(val id: Int, val name: String)